content/academy/puppeteer_playwright/collecting_data.md

```diff
@@ -0,0 +1,311 @@
+
+```shell
+npm install cheerio
+```
+
+Then, we'll import the `load` function like so:
+
+```JavaScript
+import { load } from 'cheerio';
+```
+
+Finally, we can create a `Cheerio` object based on our page's current content like so:
+
+```JavaScript
+const $ = load(await page.content());
+```
+
+> It's important to note that this `$` object is static. If any content on the page changes, the `$` variable will not automatically be updated. It will need to be redeclared or redefined.
+
+Here's our full code so far:
+
+```marked-tabs
+<marked-tab header="Playwright" lang="javascript">
+import { chromium } from 'playwright';
+import { load } from 'cheerio';
+
+const autoScroll = async (page) => {
+    await page.evaluate(async () => {
+        await new Promise((resolve) => {
+            let totalHeight = 0;
+            let distance = 100;
+            let interval = setInterval(() => {
+                const scrollHeight = document.body.scrollHeight;
+
+                window.scrollBy(0, distance);
+                totalHeight += distance;
+
+                if (totalHeight >= scrollHeight - window.innerHeight) {
+                    clearInterval(interval);
+                    resolve();
+                }
+            });
+        });
+    }, 200);
+};
+
+const browser = await chromium.launch({ headless: false });
+const page = await browser.newPage();
+
+await page.goto('https://demo-webstore.apify.org/search/on-sale');
+
+await page.waitForLoadState('networkidle');
+
+await autoScroll(page);
+
+const $ = load(await page.content());
+
+// collection code will go here
+
+await browser.close();
+</marked-tab>
+<marked-tab header="Puppeteeer" lang="javascript">
+import puppeteer from 'puppeteer';
+import { load } from 'cheerio';
+
+const autoScroll = async (page) => {
+    await page.evaluate(async () => {
+        await new Promise((resolve) => {
+            let totalHeight = 0;
+            let distance = 100;
+            let interval = setInterval(() => {
+                const scrollHeight = document.body.scrollHeight;
+
+                window.scrollBy(0, distance);
+                totalHeight += distance;
+
+                if (totalHeight >= scrollHeight - window.innerHeight) {
+                    clearInterval(interval);
+                    resolve();
+                }
+            });
+        });
+    }, 200);
+};
+
+const browser = await puppeteer.launch({ headless: false });
+const page = await browser.newPage();
+
+await page.goto('https://demo-webstore.apify.org/search/on-sale');
+
+await page.waitForNetworkIdle();
+
+await autoScroll(page);
+
+const $ = load(await page.content());
+
+// collection code will go here
+
+await browser.close();
+</marked-tab>
+```
+
+Now, to loop through all of the products, we'll make use of the `$` object and loop through them while safely in the server-side context rather than running the code in the browser.
```

Comment:
I fail to understand how the server side context is safer than the browser context. An unhandled exception in any of them will crash the program :)